### PR TITLE
Add Fallback Provider #421

### PR DESCRIPTION
**Pull request by jaredkerim**
- Add FallbackProvider
- Add allow_fallback to ApiKey
- Add fallback parameters to ichnaea.ini
- Add should_locate method to provider/searcher contract
- Add tests

### CODE REVIEW

File: ichnaea/locate/provider.py [hunk 1]
```diff
@@ -531,3 +545,63 @@ class GeoIPPositionProvider(BaseGeoIPProvider):
...
+
+    def locate(self, data):
+        location = self.location_type(query_data=False)
+
+        cell_queries = [{
+            'cellId': cell.get('cid', None),
+            'locationAreaCode': cell.get('lac', None),
+            'mobileCountryCode': cell.get('mcc', None),
+            'mobileNetworkCode': cell.get('mnc', None),
+        } for cell in data.get('cell', [])]
```
Comment (hannosch) [inline ichnaea/locate/provider.py]: This is missing the radioType field.

File: ichnaea/locate/provider.py [hunk 2]
```diff
@@ -531,3 +545,63 @@ class GeoIPPositionProvider(BaseGeoIPProvider):
...
+        wifi_queries = [{
+            'macAddress': re.sub(
+                '(..(?!$))', '\g<0>:', wifi.get('key', None)),
```
Comment (hannosch) [inline ichnaea/locate/provider.py]: Do we need to add these colons? Or does it also work without adding them?

File: ichnaea/locate/provider.py (hunk 1 above)
Comment (hannosch) [inline ichnaea/locate/provider.py]: For radio, make sure we send a plain text string, and not one of our enum's. We only need to send one radioType field in each cell entry, nothing at the top level.

File: ichnaea/locate/provider.py [hunk 3]
```diff
@@ -531,3 +545,63 @@ class GeoIPPositionProvider(BaseGeoIPProvider):
...
+        try:
+            response = requests.post(
+                self.fallback_url, json=query_data)
+        except requests.exceptions.RequestException:
+            response = None
+
+        if response:
+            try:
+                location_data = json.loads(response.content)
+            except ValueError:
```
Comment (hannosch) [inline ichnaea/locate/provider.py]: Why not just use `response.json` - via http://docs.python-requests.org/en/latest/user/quickstart/#json-response-content

File: ichnaea/locate/provider.py [hunk 4]
```diff
@@ -531,3 +545,63 @@ class GeoIPPositionProvider(BaseGeoIPProvider):
...
+                location_data = {}
+
+            location = self.location_type(
+                lat=location_data.get('lat', None),
+                lon=location_data.get('lon', None),
+                accuracy=location_data.get('accuracy', None),
+            )
```
Comment (hannosch) [inline ichnaea/locate/provider.py]: This looks like the wrong format to me, according to the docs we should expect a nested structure like:
```
{
	"location": {
		"lat": 59.331706,
		"lng": 18.079073
	},
	"accuracy": 500
}
```

File: ichnaea/locate/searcher.py [hunk 5]
```diff
@@ -118,6 +120,9 @@ class PositionSearcher(Searcher):
         ('wifi', (
             WifiPositionProvider,
         )),
+        ('fallback', (
+            FallbackProvider,
+        )),
```
Comment (hannosch) [inline ichnaea/locate/searcher.py]: Could we reorder this list and put the geoip provider last? Would that avoid the "weak location" check inside the fallback's should_locate method?

File: ichnaea/locate/provider.py [hunk 6]
```diff
@@ -531,3 +545,63 @@ class GeoIPPositionProvider(BaseGeoIPProvider):
...
+
+        query_data = {
+            'cellTowers': cell_queries,
+            'wifiAccessPoints': wifi_queries,
+        }
+
+        try:
+            response = requests.post(
+                self.fallback_url, json=query_data)
+        except requests.exceptions.RequestException:
```
Comment (hannosch) [inline ichnaea/locate/provider.py]: Can you add an override for the user-agent header? So it becomes something like I did in the export code at https://github.com/mozilla/ichnaea/blob/master/ichnaea/data/export.py#L105

Something like user-agent: ichnaea, so we don't leak what specific version of urllib/python we are using internally.

I'd also add the other two things I did there: 1. add a timeout, for this lookup make it 5 seconds and disable ssl verification for now.

IIRC we have to do a bunch of work to get SNI support working in Python 2.6.

File: ichnaea/locate/tests/test_provider.py [hunk 7]
```diff
@@ -505,3 +512,146 @@ def test_geoip_country(self):
...
+    TestProvider = FallbackProvider
+
+    def setUp(self):
+        super(TestFallbackProvider, self).setUp()
+
+        self.provider.api_key.allow_fallback = True
+
+        requests_patcher = mock.patch('ichnaea.locate.provider.requests.post')
+        self.mock_requests = requests_patcher.start()
+        self.addCleanup(requests_patcher.stop)
```
Comment (hannosch) [inline ichnaea/locate/tests/test_provider.py]: For requests mocking, we already use the extra helper library mock_requests (https://requests-mock.readthedocs.org/en/latest/). There's a couple examples over here: https://github.com/mozilla/ichnaea/blob/master/ichnaea/data/tests/test_export.py#L139

File: ichnaea/locate/tests/test_provider.py [hunk 8]
```diff
@@ -505,3 +512,146 @@ def test_geoip_country(self):
...
+        self.mock_requests.return_value = response
+
+        cells = [{
+            'cid': cid,
+            'radio': Radio.gsm.name,
+            'mcc': GB_MCC,
+            'mnc': 1,
+            'lac': 1,
+        } for cid in range(1, 3)]
+        wifis = [{'key': '001122334455'}, {'key': '112233445566'}]
```
Comment (hannosch) [inline ichnaea/locate/tests/test_provider.py]: Can you use our factory boy test factories to get these cell/wifi keys? Someone told me to use them :))

File: ichnaea/locate/tests/test_provider.py [hunk 9]
```diff
@@ -505,3 +512,146 @@ def test_geoip_country(self):
...
+        self.mock_read_config().get_map.return_value = {
+            'fallback_url': self.fallback_url,
+        }
+
+    def test_successful_call_returns_location(self):
+        response_location = {
+            'lat': 1.0,
+            'lon': 1.0,
+            'accuracy': 100,
+        }
```
Comment (hannosch) [inline ichnaea/locate/tests/test_provider.py]: Noticed this in the actual logic, this is the wrong response format. This looks like our search response, but not the geolocate API response.

**Comment by hannosch** [comment]
One more thing: I'd change the way we specify the config settings a bit. As part of my export work, I've spent a bit of time to clean that area of the code up, so we can specify multiple config sections now. So I'd change the config file over to have something like:
```
[locate:fallback]
url = https://example.com/?key=secret
ratelimit = 
cache_timeout = 
```
Specifying the api_key separately from the url isn't needed, and your code expects it in one setting as well.

To provide the test default values, you can then set up a new section in https://github.com/mozilla/ichnaea/blob/master/ichnaea/tests/base.py#L60

These config settings get parsed as part of the startup process, over here: https://github.com/mozilla/ichnaea/blob/master/ichnaea/webapp/config.py#L28. That leads to them being available as part of the request, as `request.registry.settings`. So something like `request.registry.settings.get('locate:fallback', {}).get('url', None)` will get you the url.

I'd pass the entire settings into the searcher. As part of the provider setup (https://github.com/mozilla/ichnaea/blob/master/ichnaea/locate/searcher.py#L33), I'd let it pull out the appropriate config section (`settings.get('locate:' + provider_group, {}`) out of the general settings and pass that into each provider.

That way the fallback provider ends up taking one more settings argument, which is a plain dict which at first is just a `{'url': 'https://example.com/?key=secret'}`, while later that dict automatically gets the extra settings.

Inside the tests we can avoid mocking this setting and just explicitly pass it in. Inside the application code, we don't need to call read_config and parse the config file on each request. Should any of the other providers need config settings in the future, it's trivial to add a new config section and it magically ends up in the right place. We may want to move some of the clustering constants into such settings in the future.

**Comment by hannosch** [comment]
I started looking into what kind of metrics we want to capture here, and came up with #430 first to simplify this. Once that is done we can revisit adding new metrics for the fallback provider, probably something like a timer called `locate.fallback.lookup` around the outgoing request, and counters to track the HTTP response status codes under `locate.fallback.lookup_status.200`, `locate.fallback.lookup_status.400`, ...

Not quite sure yet what kind of quality metrics we want (for example fallback returned wifi / cell / no result - how would we determine this ... or just a histogram of returned accuracy?).

File: ichnaea/locate/searcher.py (hunk 5 above)
Comment (jaredkerim) [inline ichnaea/locate/searcher.py]: Hrmm.... this needs some further research.  I'll look at it tomorrow.

**Comment by jaredkerim** [comment]
@hannosch r? should be good to merge

File: ichnaea/locate/searcher.py (hunk 5 above)
Comment (hannosch) [inline ichnaea/locate/searcher.py]: Ok, then ignore this. The weak check is actually a sensible one. We'll probably end up changing this bit soon enough. Maybe something like "if fallback is allowed for this request, do that first", so it's going to be more of a preferred provider, or do both lookups in our own data and the fallback provider and collect some metrics around how those two results differ from each other. Not sure yet, so leave as is for now :)

**Comment by jaredkerim** [comment]
There's two methods, ._to_json() and ._to_json_value()  The one you're thinking of is ._to_json(), ._to_json_value() returns just the dict we want.  I agree it's a long ugly method call, what would be nice is a different set of serializers for the same models that serialize them to the different formats depending on which API we're talking to.  We should look into some serialization modules, maybe sqlalchemy has something for serializing models to json, but we should probably do that in a separate ticket.

**Comment by jaredkerim** [comment]
That's odd, why didn't your comments get associated with lines?  It was kind of a puzzle trying to figure out which lines you were talking about!  Anyways I think that's everything, r? @hannosch again?

**Comment by hannosch** [comment]
@jaredkerim I'll have a look on Monday. As for line numbers, my guess is force-pushing destroys those, as the comments are only attached to lines from the old commit hash. I still prefer to never force push during a review cycle, and only merge and rebase at the very end.

**Comment by hannosch** [comment]
And as for the `_to_json_value` method, you are right, I looked at the wrong one. But even `_to_json_value` has a problem for cell keys: It encodes the radio field as an int, and not as `.name`, so it comes out as `{'radio': 0, mcc: 123, ...}` which is want the query wants either.

To be fair in the test where you use it, it pretty much doesn't matter, as the entire logic only checks for the number of items in the cell and wifi lists, and you could put pretty much anything in there.

**Pull request merged**